File: worker/integration_test.go
package main_test

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"strings"
	"testing"

	worker "worker"
	. "worker/utils"

	"github.com/joho/godotenv"
	rd "github.com/redis/go-redis/v9"
	"github.com/stretchr/testify/assert"
)

const (
	WORKER_ADDR           = "WORKER_ADDR"
	WORKER_PORT           = "WORKER_PORT"
	REDIS_PORT            = "REDIS_PORT"
	REDIS_ADDR            = "REDIS_ADDR"
	REDIS_PASSWORD        = "REDIS_PASSWORD"
	REDIS_WORK_PREFIX     = "REDIS_WORK_PREFIX"
	REDIS_WORK_PROCESS    = "REDIS_WORK_PROCESS"
	REDIS_WORK_RESULT     = "REDIS_WORK_RESULT"
	REDIS_WORK_STATUS     = "REDIS_WORK_STATUS"
	REDIS_CH_WORK_PROCESS = "REDIS_CH_WORK_PROCESS"
	REDIS_CH_WORK_RESULT  = "REDIS_CH_WORK_RESULT"
)

func initEnv(t *testing.T) {
	err := godotenv.Load("../.env.public", "../.env")
	assert.Nil(t, err, "Failed to load ../.env.public or ../.env environment file: %v", err)

	val, err := TryGetEnv(WORKER_ADDR)
	assert.Nil(t, err, err)
	assert.NotEqual(t, strings.Trim(val, " "), "")

	val, err = TryGetEnv(WORKER_PORT)
	assert.Nil(t, err, err)
	assert.NotEqual(t, strings.Trim(val, " "), "")

	val, err = TryGetEnv(REDIS_PORT)
	assert.Nil(t, err, err)
	assert.NotEqual(t, strings.Trim(val, " "), "")

	val, err = TryGetEnv(REDIS_ADDR)
	assert.Nil(t, err, err)
	assert.NotEqual(t, strings.Trim(val, " "), "")

	val, err = TryGetEnv(REDIS_WORK_PREFIX)
	assert.Nil(t, err, err)
	assert.NotEqual(t, strings.Trim(val, " "), "")

	val, err = TryGetEnv(REDIS_WORK_PROCESS)
	assert.Nil(t, err, err)
	assert.NotEqual(t, strings.Trim(val, " "), "")

	val, err = TryGetEnv(REDIS_WORK_RESULT)
	assert.Nil(t, err, err)
	assert.NotEqual(t, strings.Trim(val, " "), "")

	val, err = TryGetEnv(REDIS_WORK_STATUS)
	assert.Nil(t, err, err)
	assert.NotEqual(t, strings.Trim(val, " "), "")

	val, err = TryGetEnv(REDIS_CH_WORK_PROCESS)
	assert.Nil(t, err, err)
	assert.NotEqual(t, strings.Trim(val, " "), "")

	val, err = TryGetEnv(REDIS_CH_WORK_RESULT)
	assert.Nil(t, err, err)
	assert.NotEqual(t, strings.Trim(val, " "), "")
}

func fetchJobIndex(t *testing.T, httpclient *http.Client) string {
	url := fmt.Sprintf("http://%s:%s/job-index", GetEnv(WORKER_ADDR), GetEnv(WORKER_PORT))
	resp, err := httpclient.Get(url)
	assert.Nil(t, err, fmt.Sprintf("Cannot access the worker REST API /job-idx endpoint %s", url))
	defer resp.Body.Close()

	assert.Equal(t, resp.StatusCode, 200, "/job-index response status should be 200")

	body, err := io.ReadAll(resp.Body)
	assert.Nil(t, err, fmt.Sprintf("Cannot parse the /job-idx response body %s", err))
	assert.NotEmpty(t, body, "/job-index shouldn't be empty")

	return string(body)
}

func TestWorker(t *testing.T) {
	initEnv(t)

	ctx := context.Background()
	httpclient := http.Client{}
	redis := worker.RedisDefault()

	job_idx := fetchJobIndex(t, &httpclient)

	redis.Set(rd.NewClient(&rd.Options{
		Addr:     fmt.Sprintf("%s:%s", GetEnv(REDIS_ADDR), GetEnv(REDIS_PORT)),
		Password: GetEnv(REDIS_PASSWORD),
	}))

	client := redis.Get()
	defer redis.Unlock()

	config_key := fmt.Sprintf("%s:%s:%s", GetEnv(REDIS_WORK_PREFIX), job_idx, GetEnv(REDIS_WORK_PROCESS))
	err := client.HSet(ctx, config_key, "endpoint", "/api", "parameters", "classes").Err()
	assert.Nil(t, err, err)

	sub := client.Subscribe(ctx, GetEnv(REDIS_CH_WORK_RESULT))
	defer sub.Close()

	sub_mess, err := sub.Receive(ctx)
	assert.Nil(t, err, "Failed to receive message from subscriber channel (Subscription step)")

	err = client.Publish(ctx, GetEnv(REDIS_CH_WORK_PROCESS), job_idx).Err()
	assert.Nil(t, err, err)

	var result_key string

	// This might fail since it's not concurrently subscribed to the channel before triggering the work
	switch sub_mess.(type) {
	case *rd.Subscription:
		sub_mess, err = sub.Receive(ctx)
		assert.Nil(t, err, "Failed to receive message from subscriber channel (Message step)")

		switch sub_mess.(type) {
		case *rd.Message:
			result_key = sub_mess.(*rd.Message).Payload
		default:
			t.Log("Subscription channel: Failed to receive Message response")
			t.Fail()
			return
		}

	default:
		t.Log("Subscription channel: Failed to receive Subscription response")
		t.Fail()
		return
	}

	result, err := client.Get(ctx, result_key).Result()
	assert.Nil(t, err, err)

	assert.NotEqual(t, len(result), 0, "Work result string shouldn't be empty")

	// For visual debug
	// fmt.Println(result)
}

File: backend/integration_test.ts
import { assert, assertNotEquals } from "jsr:@std/assert";
import process from "node:process";
import dotenv, { type DotenvPopulateInput } from "npm:dotenv@16.4.5";
import { connect as createRedisClient, type RedisConnectOptions } from "https://deno.land/x/redis@v0.33.0/mod.ts";

const ENV_VARS = {
	WORKER_ADDR: "WORKER_ADDR",
	WORKER_PORT: "WORKER_PORT",
	REDIS_PORT: "REDIS_PORT",
	REDIS_ADDR: "REDIS_ADDR",
	REDIS_PASSWORD: "REDIS_PASSWORD",
	REDIS_WORK_PREFIX: "REDIS_WORK_PREFIX",
	REDIS_WORK_PROCESS: "REDIS_WORK_PROCESS",
	REDIS_WORK_RESULT: "REDIS_WORK_RESULT",
	REDIS_WORK_STATUS: "REDIS_WORK_STATUS",
	REDIS_CH_WORK_PROCESS: "REDIS_CH_WORK_PROCESS",
	REDIS_CH_WORK_RESULT: "REDIS_CH_WORK_RESULT",
};

// Resource sanitize disabled since I can't, for some reason, close/shutdown the Redis TCP connection(s)
Deno.test("Worker input to output", { sanitizeResources: false }, async t => {
	t.step("Configuring ENV based on .env* files", () => {
		for (const path of ["../.env", "../.env.public"]) {
			const err = dotenv.config({ path, processEnv: process.env as DotenvPopulateInput }).error;
			if (err != undefined) {
				throw err;
			}
		}

		for (const env_var in ENV_VARS) {
			assertNotEquals(process.env[env_var], undefined);
			assertNotEquals(process.env[env_var], "");
		}
	});

	const redis_options: RedisConnectOptions = {
		hostname: process.env[ENV_VARS.REDIS_ADDR]!,
		port: process.env[ENV_VARS.REDIS_PORT],
		password: process.env[ENV_VARS.REDIS_PASSWORD],
	};

	const client = await createRedisClient(redis_options);
	const sub_client = await createRedisClient(redis_options);

	await client.connect();
	await sub_client.connect();

	const resp = await fetch(
		`http://${process.env[ENV_VARS.WORKER_ADDR]}:${process.env[ENV_VARS.WORKER_PORT]}/job-index`,
		{
			method: "GET",
		},
	);
	const job_idx = await resp.text();

	assertNotEquals(job_idx.length, 0, "/job-index response shouldn't be empty");

	const config_key = `${process.env[ENV_VARS.REDIS_WORK_PREFIX]}:${job_idx}:${process.env[ENV_VARS.REDIS_WORK_PROCESS]}`;
	await client.hset(config_key, {
		endpoint: "/api/classes",
		parameters: "warlock",
	});

	await client.publish(process.env[ENV_VARS.REDIS_CH_WORK_PROCESS]!, job_idx);

	const sub = await sub_client.subscribe(process.env[ENV_VARS.REDIS_CH_WORK_RESULT]!);

	let result_key = "";
	for await (const { message } of sub.receive()) {
		result_key = message;
		break;
	}

	sub.close();

	const result = await client.get(result_key);

	assertNotEquals(result, null);
	assertNotEquals(result!.trim().length, 0);
	assertNotEquals(result, "nil");

	client.close();
	sub_client.close();

	assert(client.isClosed);
	assert(sub_client.isClosed);

	// console.log(result);
});
